gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/extractor/FileAwareInputStreamExtractorWithCheckSchema.java

```diff
@@ -39,32 +42,109 @@
-      throws IOException {
+  protected boolean schemaChecking(FileSystem fsFromFile) throws IOException {
     DatumReader<GenericRecord> datumReader = new GenericDatumReader<>();
-    DataFileReader<GenericRecord> dataFileReader = new DataFileReader(new FsInput(this.file.getFileStatus().getPath(),fsFromFile), datumReader);
+    DataFileReader<GenericRecord> dataFileReader =
+        new DataFileReader(new FsInput(this.file.getFileStatus().getPath(), fsFromFile), datumReader);
     Schema schema = dataFileReader.getSchema();
-    return schema.toString().equals(this.state.getProp(ConfigurationKeys.COPY_EXPECTED_SCHEMA));
+    Schema expectedSchema = new Schema.Parser().parse(this.state.getProp(ConfigurationKeys.COPY_EXPECTED_SCHEMA));
+
+    return compare(schema, expectedSchema);
+  }
+
+  private boolean compare(Schema toValidate, Schema expected) {
+    if (toValidate.getType() != expected.getType() || !toValidate.getName().equals(expected.getName())) {return false;}
+    else {
+      switch (toValidate.getType()) {
+        case NULL:
+        case BOOLEAN:
+        case INT:
+        case LONG:
+        case FLOAT:
+        case DOUBLE:
+        case BYTES:
+        case STRING: {
+          return true;
+        }
+        case ARRAY: {
+          return compare(toValidate.getElementType(), expected.getElementType());
+        }
+        case MAP: {
+          return compare(toValidate.getValueType(), expected.getValueType());
+        }
+        case FIXED: {
+          // fixed size and name must match:
+          if (toValidate.getFixedSize() != expected.getFixedSize()) {
+            return false;
+          }
+        }
+        case ENUM: {
+          // expected symbols must contain all toValidate symbols:
+          final Set<String> expectedSymbols = new HashSet<String>(expected.getEnumSymbols());
+          final Set<String> toValidateSymbols = new HashSet<String>(toValidate.getEnumSymbols());
+          if (expectedSymbols.size() != toValidateSymbols.size()) {
+            return false;
+          }
+          if (!expectedSymbols.containsAll(toValidateSymbols)) {
+            return false;
+          }
+        }
+
+        case RECORD: {
+          // Check that each field of toValidate schema is in expected schema
+          if(toValidate.getFields().size() != expected.getFields().size()) {return false;}
+          for (final Schema.Field expectedFiled : expected.getFields()) {
+            final Schema.Field toValidateField = toValidate.getField(expectedFiled.name());
+            if (toValidateField == null) {
+              // expected field does not correspond to any field in the toValidate record schema
+              return false;
+            } else {
+              if (!compare(toValidateField.schema(), expectedFiled.schema())) {
+                return false;
+              }
+            }
+          }
+          return true;
+        }
+        case UNION: {
```

Comment:
Just want to confirm, do you care about the ordering of fields ?

Reply:
No, I just want to make sure they have matching field.